### PR TITLE
JAX Linear Regression Example - Interactive API

### PR DESCRIPTION
**Pull request by sun1lach**

### CODE REVIEW

**Comment by sun1lach** [comment]
I have read the CLA Document and I hereby sign the CLA

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,133 @@
+absl-py==1.1.0
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/jax_linear_regression/workspace/requirements.txt]: I believe the `requirements.txt` is auto-generated by OpenFL.

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Copyright (C) 2020-2021 Intel Corporation
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py]: 2022

File: openfl-tutorials/interactive_api/jax_linear_regression/envoy/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,2 @@
+openfl==1.3
+numpy
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/jax_linear_regression/envoy/requirements.txt]: `jax`?

File: openfl-tutorials/interactive_api/jax_linear_regression/envoy/envoy_config_2.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+params:
+  cuda_devices: []
+
+optional_plugin_components: {}
+
+shard_descriptor:
+  template: linreg_shard_descriptor.LinRegSD
+  params:
+    rank: 43 # Dataset seed
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/jax_linear_regression/envoy/envoy_config_2.yaml]: `rank` might confuse users. It is used in the context of sharding across participants.
You may want to encode these parameters within the notebook instead of the envoy config. This config typically should hold federation-related parameters only.

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+# SPDX-License-Identifier: Apache-2.0
+"""Custom model numpy adapter."""
+
+from openfl.plugins.frameworks_adapters.framework_adapter_interface import (
+    FrameworkAdapterPluginInterface,
+)
+
+
+class CustomFrameworkAdapter(FrameworkAdapterPluginInterface):
+    """Framework adapter plugin class."""
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py:11]: Is this the subclassing needed for JAX or numpy?

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+    @staticmethod
+    def get_tensor_dict(model, optimizer=None):
+        """Extract tensors from a model."""
+        return {'w': model.weights}
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py:16]: Will this work for all generic `jax` models or just this regression example?

File: openfl-tutorials/interactive_api/jax_linear_regression/envoy/linreg_shard_descriptor.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        Return a shard dataset by type.
+
+        A simple list with elements (x, y) implements the Shard Dataset interface.
+        """
+        if dataset_type == 'train':
+            return self.train
+        elif dataset_type == 'val':
+            return self.test
+        else:
+            return None
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/jax_linear_regression/envoy/linreg_shard_descriptor.py]: `else` return `ValueError`?

**Review by MasterSkepticista — CHANGES_REQUESTED** [review]

File: openfl-tutorials/interactive_api/jax_linear_regression/README.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+```sh
+./start_envoy.sh env_instance_2 envoy_config_2.yaml
+```
+
+3. Run `jax_linear_regression.ipybnb` jupyter notebook:
+
+```sh
+cd workspace
+jupyter lab Mnist_Classification_FL.ipybnb
```
Comment (mansishr) [inline openfl-tutorials/interactive_api/jax_linear_regression/README.md]: 'ipynb' typo

File: openfl-tutorials/interactive_api/jax_linear_regression/director/start_director_with_tls.sh [hunk 9]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+set -e
+FQDN=$1
+fx director start -c director_config.yaml -rc cert/root_ca.crt -pk cert/"${FQDN}".key -oc cert/"${FQDN}".crt
```
Comment (mansishr) [inline openfl-tutorials/interactive_api/jax_linear_regression/director/start_director_with_tls.sh]: Let's either add how to run the tutorial with TLS in README or remove this file

File: openfl-tutorials/interactive_api/jax_linear_regression/README.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
+# JAX based Linear Regression Tutorial
+
+### 1. About dataset
+
+Generate a random regression problem using `make_regression` from sklearn.datasets.
+Define the below params in envoy.yaml config to control the dataset generation.
+- rank (seed)
```
Comment (mansishr) [inline openfl-tutorials/interactive_api/jax_linear_regression/README.md]: Let's rename rank to avoid confusion with shard descriptor rank parameter.

**Review by mansishr — CHANGES_REQUESTED** [review]

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/requirements.txt [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+matplotlib-inline==0.1.3
+mistune==0.8.4
+nbclassic==0.3.7
+nbclient==0.6.4
+nbconvert==6.5.0
+nbformat==5.4.0
+nest-asyncio==1.5.5
+notebook==6.4.12
+notebook-shim==0.1.0
+numpy==1.22.4
```
Comment (shreyasnc1) [inline openfl-tutorials/interactive_api/jax_linear_regression/workspace/requirements.txt]: Required?

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/JAX_linear_regression.ipynb [hunk 12]
```diff
@@ -0,0 +1,561 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "bcb812db-971e-4b25-83a8-86bfa901af11",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "#### Linear Regression Model without federeation"
```
Comment (shreyasnc1) [inline openfl-tutorials/interactive_api/jax_linear_regression/workspace/JAX_linear_regression.ipynb]: Is this required?

File: openfl-tutorials/interactive_api/jax_linear_regression/envoy/start_envoy_with_tls.sh [hunk 13]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+set -e
+ENVOY_NAME=$1
+ENVOY_CONF=$2
+DIRECTOR_FQDN=$3
+
+fx envoy start -n "$ENVOY_NAME" --envoy-config-path "$ENVOY_CONF" -dh "$DIRECTOR_FQDN" -dp 50050 -rc cert/root_ca.crt -pk cert/"$ENVOY_NAME".key -oc cert/"$ENVOY_NAME".crt
```
Comment (shreyasnc1) [inline openfl-tutorials/interactive_api/jax_linear_regression/envoy/start_envoy_with_tls.sh]: Is this tested?

File: openfl-tutorials/interactive_api/jax_linear_regression/envoy/linreg_shard_descriptor.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
+# Copyright (C) 2020-2021 Intel Corporation
```
Comment (shreyasnc1) [inline openfl-tutorials/interactive_api/jax_linear_regression/envoy/linreg_shard_descriptor.py]: 2022 here and everywhere else

File: openfl-tutorials/interactive_api/jax_linear_regression/envoy/envoy_config_2.yaml [hunk 15]
```diff
@@ -0,0 +1,13 @@
+params:
```
Comment (shreyasnc1) [inline openfl-tutorials/interactive_api/jax_linear_regression/envoy/envoy_config_2.yaml]: The second envoy confg is not required. You have already mentioned in the README on how to use the config for other envoys

**Review by shreyasnc1 — CHANGES_REQUESTED** [review]

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/requirements.txt (hunk 1 above)
Comment (sun1lach) [reply openfl-tutorials/interactive_api/jax_linear_regression/workspace/requirements.txt]: Sure. 
Refactored example to run independent of `requirements.txt` package dependency.

File: openfl-tutorials/interactive_api/jax_linear_regression/envoy/requirements.txt (hunk 3 above)
Comment (sun1lach) [reply openfl-tutorials/interactive_api/jax_linear_regression/envoy/requirements.txt]: Is it expected to have all the relevant packages pre-installed in the envoy? or OpenFL picks up the necessary packages to install from `requirements.txt` present in envoy directory?

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py (hunk 5 above)
Comment (sun1lach) [reply openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py:11]: `get_tensor_dict` and `set_tensor_dict` method is overridden in this custom adapter class to load and set weights. 
Required at the time when collaborator returns weights for aggregation.
Although the adapters are defined in `openfl.plugins.frameworks_adapters`, custom adapters are defined in `interactive_api` examples. Same is the case for other examples as well.
Need to streamline this?

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py (hunk 6 above)
Comment (sun1lach) [reply openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py:16]: just for this regression example.

File: openfl-tutorials/interactive_api/jax_linear_regression/README.md (hunk 8 above)
Comment (sun1lach) [reply openfl-tutorials/interactive_api/jax_linear_regression/README.md]: Corrected.

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/requirements.txt (hunk 11 above)
Comment (sun1lach) [reply openfl-tutorials/interactive_api/jax_linear_regression/workspace/requirements.txt]: Removed the `requirements.txt` from workspace.
Added jax and jaxlib pip install directly in notebook.

File: openfl-tutorials/interactive_api/jax_linear_regression/envoy/start_envoy_with_tls.sh (hunk 13 above)
Comment (sun1lach) [reply openfl-tutorials/interactive_api/jax_linear_regression/envoy/start_envoy_with_tls.sh]: I will remove the TLS part for now.

File: openfl-tutorials/interactive_api/jax_linear_regression/envoy/requirements.txt (hunk 3 above)
Comment (MasterSkepticista) [reply openfl-tutorials/interactive_api/jax_linear_regression/envoy/requirements.txt]: The `requirements.txt` within envoy directory installs the specific example related packages in general. Here, it should be numpy, jax, etc. If it was a TF example, `tensorflow` would be a required package for the example.

That is my understanding.

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py (hunk 5 above)
Comment (MasterSkepticista) [reply openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py:11]: My primary concern is, that this subclassing should work for all JAX models and not just linear regression.

Since you're adding a JAX example, your framework adapter should be able to handle other model architectures as well. For instance, a `pytorch` `model.state_dict()` can be iteratively mapped for each layer regardless of the architecture.

I'll let other reviewers chime in if this needs to be enforced or if they would like to archive it for future discussions. If latter, I'm ok and this can be closed.

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py (hunk 6 above)
Comment (MasterSkepticista) [reply openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py:16]: Since you're adding a JAX example, your framework adapter should be able to handle other model architectures as well. For instance, a `pytorch` `model.state_dict()` can be iteratively mapped for each layer regardless of the architecture.

I'll let other reviewers chime in if this needs to be enforced or if they would like to archive it for future discussions. If latter, I'm ok and this can be closed.

File: openfl-tutorials/interactive_api/jax_linear_regression/envoy/envoy_config_2.yaml (hunk 4 above)
Comment (shreyasnc1) [reply openfl-tutorials/interactive_api/jax_linear_regression/envoy/envoy_config_2.yaml]: @MasterSkepticista  Not sure why you think this should hold only federation related config only. From the examples I see the SD related configs are defined here. The shard descriptor is created from here, so parameterizing is not possible. Am I missing something?

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py (hunk 6 above)
Comment (shreyasnc1) [reply openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py:16]: @acharyasunil  this should work for all JAX models. @MasterSkepticista  What is the rationale/usecase behind " your framework adapter should be able to handle other model architectures as well" ?  I didnt get the usecase behind pytorch model.state_dict()
Eventually tensor dicts are converted to numpy before sending to aggregator.

File: openfl-tutorials/interactive_api/jax_linear_regression/envoy/envoy_config_2.yaml (hunk 4 above)
Comment (MasterSkepticista) [reply openfl-tutorials/interactive_api/jax_linear_regression/envoy/envoy_config_2.yaml]: This is the only way of parameterizing the `ShardDescriptor`.
Args like `rank`, `img_size`, `transform`, or `data_path` can make it to the `params`, because they're relevant to the federation tasks of collaborators in general. Bottom 3 args in this config are for generating any random data and should be better done in the experiment instead of controlling via `envoy_config`

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/JAX_linear_regression.ipynb (hunk 12 above)
Comment (sun1lach) [reply openfl-tutorials/interactive_api/jax_linear_regression/workspace/JAX_linear_regression.ipynb]: Made it optional at the end of notebook.

File: openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py (hunk 6 above)
Comment (sun1lach) [reply openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py:16]: Still exploring this part. Will keep this conversation open.

Comment (sun1lach) [reply openfl-tutorials/interactive_api/jax_linear_regression/workspace/custom_adapter.py:16]: It would be better to have a structured `get_tensor_dict` and `set_tensor_dict` with optimizer state for JAX based Neural Network example.
For the current regression example, a plain getter and setter for model.weights applicable for all JAX model is sufficient.

**Review by shreyasnc1 — APPROVED** [review]

**Review by MasterSkepticista — APPROVED** [review]

**Review by psfoley — APPROVED** [review]
